src/ducc0/infra/threading.h

```diff
@@ -53,21 +53,106 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #ifndef DUCC0_THREADING_H
 #define DUCC0_THREADING_H
 
+// Low level threading support can be influenced by the following macros:
+// - DUCC0_NO_LOWLEVEL_THREADING: if defined, multithreading is disabled
+//   and all parallel regions will be executed sequentially
+//   on the invoking thread.
+// - DUCC0_CUSTOM_LOWLEVEL_THREADING: if defined, external definitions of
+//   Mutex, UniqueLock, LockGuard, CondVar, thread_pool, set_active_pool(),
+//   and get active_pool() must be supplied,
+//   and the code will use those.
+// Both macros must not be defined at the same time.
+// If neither macro is defined, standard ducc0 multihreading will be active.
+
+#if (defined(DUCC0_NO_LOWLEVEL_THREADING) && defined(DUCC0_CUSTOM_LOWLEVEL_THREADING))
+static_assert(false, "DUCC0_NO_LOWLEVEL_THREADING and DUCC0_CUSTOMLOWLEVEL_THREADING must not be both defined");
+#endif
+
+#if defined(DUCC0_STDCXX_LOWLEVEL_THREADING)
+static_assert(false, "DUCC0_STDCXX_LOWLEVEL_THREADING must not be defined externally");
+#endif
+
+#if ((!defined(DUCC0_NO_LOWLEVEL_THREADING)) && (!defined(DUCC0_CUSTOM_LOWLEVEL_THREADING)))
+#define DUCC0_STDCXX_LOWLEVEL_THREADING
+#endif
+
 #include <cstddef>
 #include <functional>
-#ifndef DUCC0_NO_THREADING
+#include <optional>
+
+// threading-specific headers
+#ifdef DUCC0_STDCXX_LOWLEVEL_THREADING
 #include <mutex>
 #include <condition_variable>
 #endif
-#include <optional>
+
+#ifdef DUCC0_NO_LOWLEVEL_THREADING
+// no headers needed
+#endif
+
+#ifdef DUCC0_CUSTOM_LOWLEVEL_THREADING
+// include headers necessary for custom threading support here.
+#endif
 
 namespace ducc0 {
 
 namespace detail_threading {
 
+class thread_pool;
```

Comment:
This being a non-abstract class might be somewhat limiting - forcing there to be a single definition across all linked modules.  This will prevent easily swapping in different threadpool implementations (for example, I can no longer use both Eigen's and TF's threadpools within the same program).

Reply:
Seems that I didn't anticipate how ambitious your plans for this were - sorry for the back-and forth :-) I'll revert this to abstract class.

I expect I'll have to put the `thread_pool` definition before the `#include "ducc0_custom_lowlevel_threading.h"` then, correct?